src/Xhgui/Saver/File.php

```diff
@@ -1,17 +1,95 @@
 <?php
 
+/**
+ * File saving handler
+ */
 class Xhgui_Saver_File implements Xhgui_Saver_Interface
 {
-    private $_file;
+    /**
+     * @var string
+     */
+    private $file;
 
-    public function __construct($file)
+    /**
+     * @var string
+     */
+    private $path;
+
+    /**
+     * @var bool
+     */
+    private $separateMeta;
+
+    /**
+     * Xhgui_Saver_File constructor.
+     * @param string $path
+     * @param string $file or null for default
+     * @param bool $separateMeta
+     */
+    public function __construct($path, $file, $separateMeta = false)
     {
-        $this->_file = $file;
+        $this->path             = rtrim($path, '/\\').DIRECTORY_SEPARATOR;
+        $this->file             = !empty($file) ? $file : self::getFilename();
+        $this->separateMeta     = $separateMeta;
     }
 
+    /**
+     * @param array $data
+     * @return bool|int
+     */
     public function save(array $data)
     {
-        $json = json_encode($data);
-        return file_put_contents($this->_file, $json.PHP_EOL, FILE_APPEND);
+        if ($this->separateMeta) {
+            $profiles           = Xhgui_Util::getDataForStorage($data['profile'], true);
+
+            $meta               = $data['meta'];
+
+            // store summary in separate meta file to speed up aggregation
+            $meta['summary']    = $data['profile']['main()'];
+            $meta               = Xhgui_Util::getDataForStorage($meta, false);
+
+            file_put_contents($this->path.$this->file.'.meta',$meta.PHP_EOL, FILE_APPEND);
+
+            return file_put_contents($this->path.$this->file,$profiles.PHP_EOL, FILE_APPEND);
+        }
+
+        $json = Xhgui_Util::getDataForStorage($data);
+        return file_put_contents($this->path.$this->file, $json.PHP_EOL, FILE_APPEND);
+    }
+
+    /**
+     * Get filename to use to store data
+     * @param string $dir
+     * @return string
+     */
+    public static function getFilename() {
+
+        $fileNamePattern = '';
+
+        $prefix = 'xhgui.data.'.microtime(true);
+        
+        if (empty($_SERVER['REQUEST_URI'])) {
+            if (version_compare(PHP_VERSION, '7.0.0') >= 0) {
+                try {
+                    $fileNamePattern = $prefix.bin2hex(random_bytes(5));
```

Comment:
not sure about introducing such complex logic to this project. there exists [paragonie/random_compat](https://github.com/paragonie/random_compat) for this.

is such new dependency ok @markstory?

is such high randomness really needed @GrzegorzDrozd?

Reply:
This would introduce another dependency that is php version dependant - there is no way to tell composer to install that package only for version lower than 7.
As for complexity - I am using xhgui in high traffic environment and I had few times where generated id was duplicated.

Reply:
yes, there is a way to tell composer exactly that. see random compat readme.

Reply:
@markstory are you ok adding `paragonie/random_compat` dependency? I'll add it as separate PR, so this branch can just use `random_*` family functions.

ps: I'm not fine dropping PHP 5.3 myself, I still need it, and it's not maintenance burden to keep that support.

Reply:
@markstory wait. `paragonie/random_compat` is already here:  #11

@GrzegorzDrozd pelase use `random_*` family functions

Reply:
I will change that to use random_*. Not sure how I missed it when I was syncing that with upstream.